Match ergonomics 2024

Jules-Bertholet opened the pull request:
[Rendered](https://github.com/Jules-Bertholet/rfcs/blob/match-ergonomics-2024/text/3627-match-ergonomics-2024.md)

Changes to match ergonomics for the 2024 edition.

@rustbot label T-lang A-patterns A-edition-2024

Tracking:

- https://github.com/rust-lang/rust/issues/123076

kennytm commented on text/3627-match-ergonomics-2024.md:
````diff
@@ -0,0 +1,337 @@
+inherited reference, consuming it. A pattern that does this has no other effect.
+
+```rust
+// New in edition 2023:
````
```suggestion
// New in edition 2024:
```

kennytm commented on text/3627-match-ergonomics-2024.md:
````diff
@@ -0,0 +1,337 @@
+(including "inherited" references).
+
+```rust
+let &foo = &mut 42;
````
`foo` is still immutable right?

```rust
let &foo = &mut 42;
foo = 53; // error[E0384]: cannot assign twice to immutable variable `foo`
```

Jules-Bertholet replied:
Yes. (It would be immutable with `let &mut foo = &mut 42;` also.)

kennytm commented on text/3627-match-ergonomics-2024.md:
````diff
@@ -0,0 +1,337 @@
+let [mut a] = &[0u8]; // `a` is `u8` on edition ≤ 2021, but `&u8` on edition ≥ 2024
+```
+
+Instances of such incompatibilities appear to be common, but far from unknown
````
```suggestion
Instances of such incompatibilities appear to be uncommon, but far from unknown
```
?

rpjohnst commented:
The "`&` patterns matching against `&mut`" part seems reasonable but also fairly independent of the other changes. It might make more sense as part of a deref patterns RFC, which might (or might not!) also choose to use `&` more generally.

Jules-Bertholet commented:
> The "`&` patterns matching against `&mut`" part seems reasonable but also fairly independent of the other changes.

It could be split off, yes. I mention this in the rationale section, but one of the motivating factors for including it is the ["no inherited `ref mut` behind `&`"](https://github.com/Jules-Bertholet/rfcs/blob/match-ergonomics-2024/text/3627-match-ergonomics-2024.md#all-editions-the-default-binding-mode-is-never-set-to-ref-mut-behind-an--pattern-or-reference) rule. With that rule, whether an inherited reference is `ref` or `ref mut` (and therefore, whether it can be matched with `&mut`) depends on non-local factors, so allowing users to always use `&` unless they need mutation becomes especially important.

rpjohnst commented:
The "no inherited `ref mut` behind `&`" rules seems equally independent and forwards-compatible. (And also equally reasonable! Just probably worth considering holistically alongside deref patterns.)

Jules-Bertholet commented:
> The "no inherited `ref mut` behind `&`" rules seems equally independent and forwards-compatible.

It is *not*. On edition 2024, this example would work without the rule, but errors with it: `let &[[&mut a]] = &[&mut [42]];`

(I'll edit the RFC to make that more clear)

rpjohnst commented:
I'm not sure why (a version of) "no inherited `ref mut` behind `&`" couldn't accept that example. The binding mode could be something like "`ref mut` behind `ref`" rather than flattening all the way to `ref`. Indeed, places behave similarly: given `x: &&mut i32`, `*x` still has type `&mut i32`. You may not be able to move or write through it, but that doesn't make it a `&i32`.

Really I'm not sure why we would even want to reject it in the first place. There *is* still a `&mut` there, after all, and the problem only arises if we try to move it out from behind a `&`. It would be unfortunate if, for example, `let [[&mut a]] = [&mut [42]]` worked but `let &[[&mut a]] = &[&mut [42]]` did not.

This is exactly the sort of thing that would be good to work out as part of deref patterns. Going back to the comparison with places, the reason you can get a `&i32` from `*x` is that auto(de)ref performs a reborrow- the exact expression-side equivalent to deref patterns.

Accepting this example is a different sort of consistency than the "immutability takes precedence" one. Instead, it's "inherited reference type matches skipped reference type." This also brings back the locality that "`&` patterns matching against `&mut`" is trying to compensate for.

If we want to defer any decision at all here, I believe we could instead forbid matching `&mut` against "`ref mut` behind `ref`" entirely. That would make it forward compatible to implement either "no inherited `ref mut` behind `&`" or "default `ref mut` bindings behind `ref` are `&`."

Jules-Bertholet commented:
> Instead, it's "inherited reference type matches skipped reference type."

If there are several skipped reference types with different mutabilities, you have to choose: eg in `let [[/*HERE*/ x]] = &[&mut[42]];`, which mutability do you use to determine that of the inherited reference at `HERE`? So it's impossible to be fully consistent with the principle you give.

jplatte commented on text/3627-match-ergonomics-2024.md:
````diff
@@ -0,0 +1,405 @@
+Therefore, it is possible to have a mutable by-reference binding. (An explicit
+syntax for this is left to a future RFC.)
+
+```rust
+//! Edition ≥ 2024
+let &[mut a] = &[42];
+a = &47;
+```
````
I had no big trouble following the RFC until this example. Why is `a` a reference here? The reason it's a mutable binding is clear, but I see a `&` on both sides of `=` so shouldn't `a` just be a plain integer?

Also, would `let [mut a] = &[42];` have the same behavior? That's what I would probably want to write to get a reference binding that's mutable.

jplatte replied:
Aha! That explains why I got confused 😄

jplatte commented on text/3627-match-ergonomics-2024.md:
````diff
@@ -0,0 +1,405 @@
+let _: &mut u8 = foo;
+```
+
+## Edition 2024: `&` and `&mut` can match against inherited references
````
I like it! With this, is there even a reason to keep supporting `&mut` in patterns?

Nadrieril replied:
Of course, without it we can't get desugar match ergonomics that get mutable references into places: `let &mut Struct { ref mut field } = ...`

jplatte replied:
Right, okay. Sounds worthy of a clippy lint for if it's not combined with `ref mut` then :)

rpjohnst commented:
> If there are several skipped reference types with different mutabilities, you have to choose: ..., which mutability do you use to determine that of the inherited reference?

The innermost one, of course! Why would it be anything else? That's how default binding modes already work, we're only exposing that to the user so they can exit the current mode.

Jules-Bertholet commented:
> The innermost one, of course! Why would it be anything else? That's how default binding modes already work

No it's not! Outer shared + inner mutable results in `ref`, not `ref mut`. https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c0c3f5fbe3499b54813bbe3af95e31eb

rpjohnst commented:
What I'm saying is that it is how *initially entering* a default binding mode already works. That's why the example `let &[[x]] = &[&mut [42]]` produces `x: &mut i32` even though it's underneath a `&`.

The inconsistency we're both talking about is the difference in behavior between "behind an explicit `&`" and "in default `ref` mode," demonstrated by the example above and your latest example `let [a] = &&mut [42]` => `a: &i32`. You're proposing we resolve it in favor of the binding mode; I'm proposing we resolve it in favor of the actual reference type, for the reasons I gave above- greater locality, and alignment with places and auto(de)ref.

Jules-Bertholet commented:
Here's another way of looking at it: if `&mut` patterns "unwrap" a mutable reference, it should be possible to get a mutable reference by removing the pattern.

Simple case:

```rust
let &mut x = &mut 42; // x: i32
let x = &mut 42; // x: &mut i32
```

Roundabout case:

```rust
let &[&mut x] = &[&mut 42]; // x: i32
//let &[x] = &[&mut 42]; // ERROR, but…
let &[ref x] = &[&mut 42]; // x: &&mut i32 -> we can get an &mut in the end
```

However:

```rust
let &[[&mut x]] = &[&mut [42]]; // If we allow this, with x: i32
//let &[[x]] = &[&mut [42]]; // and then remove the &mut… -> ERROR move check

// nothing we do will get us &mut i32, even without &
```

rpjohnst commented:
Indeed, that is a demonstration of the fact that we are limited to a single layer of inherited reference/a single binding mode. If we relaxed that (not that I think this would be useful in practice), you *could* get a `&mut` in that last example:

```rust
let &[[&mut x]] = &[&mut [42]]; // x: i32
let &[[x]] = &[&mut [42]]; // ERROR move check, unless the pattern equivalent of auto(de)ref coerces &mut i32 -> x: &i32
let &[[ref x]] = &[&mut [42]]; // HYPOTHETICAL: A `ref` binding + an inherited `&mut` -> x: &&mut i32
```

My proposal is not so much breaking the user's ability to remove a `&mut` pattern to get a mutable reference, but (somewhat uselessly, on its own) extending that ability to all inherited `&mut`s, and then restricting their use in the same way they are restricted in places. (With the additional possibility of making them nicer to work with via an auto(de)ref-like mechanism, as part of deref patterns.)

Nadrieril commented:
> that is a demonstration of the fact that we are limited to a single layer of inherited reference/a single binding mode. If we relaxed that...

It is semantically impossible to relax that. E.g.
```rust
let var = Some(0);
if let Some(x) = &&var {
   //...
}
```

Here `x` cannot be `&&u32`, because where would we store the `&u32` that it points to? There is no `&u32` in memory we could make it point to. There is however a `u32`, so we can have `x: &u32` pointing inside `var`.

rpjohnst commented:
Well, that was only a hypothetical to demonstrate a different way of thinking about the limitation Jules pointed out. But I don't think it's entirely *impossible* to materialize those extra reference layers- we could introduce temporaries for them around the match, it would just be a bit silly. (Not to mention confusing to be able to layer binding modes like my example.)

dev-ardi commented on text/3627-match-ergonomics-2024.md:
```diff
@@ -284,7 +284,7 @@ extern crate foo;
 
 fn main() {
-    let [[&x], foo!(y)] =  &[[&0], [0]];
+    let ([&x], foo!(y)) =  &([&0], [0]);
```
You might want to squash all of these commits

nikomatsakis commented:
@rfcbot fcp merge

We discussed this RFC in the lang-team design meeting. There was general excitement about the contents (subject to one minor caveat which I will file as a concern). The reasoning made sense and the changes felt like they would make the language more true to itself. Therefore, I am moving to fcp merge now.

nikomatsakis commented:
@rfcbot concern leaving-us-in-an-inconsitent-state

The RFC as proposed leaves us in a bit of inconsistent state. It means that you can write 

```rust
let [mut x] = &mut [22];
```

...and get a binding `mut x: &mut i32`, but there is no fully elaborated form of this, you would need...

```rust
let &mut [ref mut (mut x)] = &mut [22];
```

...but the RFC (intentionally) avoids proposing that syntax or any explicit replacement for it.

In the meeting we identified three options:

1. Make `mut x` in inherited patterns an error; convert `mut x` with inherited ref to `& (mut x)` or `&mut (mut x)` in Rust 2021.
    * can be extended later backwards compatibly (even in Rust 2024) with the semantics described here (and an explicit syntax)
2. Support `mut x` with inherited refs but do not have an explicit syntax. We probably want an explicit syntax later, but we don't have to discuss it now. <-- current RFC
3. Add `ref (mut x)` and `ref mut (mut x)` syntax (or some alternative) now 

Of these, we felt that option 2 (the current text) was in some ways the worse. Option 1 might be annoying but it's internally consistent and leaves room for Option 3. Option 3 is perhaps the most nice but also adds confusing syntax for what feels like an edge case.

**In the meeting we felt it was best to start with Option 1 for now and consider Option 3 later.**

We identified the following desirable properties (we are sacrificing 1 for now when there are inherited references):

1. You can change `x` to `mut x` to get mutability -- good for users! Writing `let mut x = x` is possible but surprising.
2. There is an explicit form in terms of `&` and `ref` etc for everything you can write with inherited mutability.
3. Property that, if you have a pattern P without inherited references, e.g., `let &(mut x) = &22` gives `mut x: i32`, and you delete an `&`, then the bindings below become of type `&T` (so e.g. `let mut x = &22`, now `mut x: &i32`)
    * "weak version" == if both compile, it holds

joshtriplett commented:
@rfcbot concern switch-to-option-1

I share @nikomatsakis's concern, and I would specifically like to go with option 1. This does not close the door on doing option 3 in the future, after introducing new syntax for it.

Jules-Bertholet commented:
I've updated the RFC in light of today's lang team design meeting ([minutes](https://hackmd.io/9SstshpoTP60a8-LrxsWSA?view)).

- Mutable by-reference bindings are now an error (addresses outstanding concerns)
- Added more examples to the rationale section for the "no `ref mut` behind `&`" rule that was discussed in the last few minutes of the meeting

rpjohnst commented:
I would still prefer to be able to consistently write a `&mut` pattern that matches an inherited reference, regardless of whether the binding mode has been converted to `ref` by an outer `&`, so long as there is a `&mut` value involved.

As we discussed above, this makes the set of available patterns more local, and relaxes the need to commit to `&`-matching-`&mut` in the short term.

Nadrieril commented:
This makes the model more complex, since we would need to distinguish "ref binding mode" from "ref binding mode that can be matched with `&mut`".

rpjohnst commented:
In that sense, the current RFC *also* introduces an additional binding mode: "`move` binding mode that transitions to `ref` on `&mut`." In this framing, either choice of "`move` that transitions to `ref`" or "`ref` matchable by `&mut`" adds a new distinction to match ergonomics. However, I think both of these ways of thinking are focused too narrowly on patterns and binding modes, to the detriment of the language as a whole.

Another framing, rather than changing the binding modes at all, is to defer this sort of "`&mut` -> `&`" adjustment to an auto(de)ref-like mechanism at the site of the binding. This (again) corresponds more closely to how expressions work: we leave the types of deref place expressions alone, even when moving out of them would be an error, and instead we defer that issue to the containing expression, which may trigger a coercion.

When the inherited reference corresponding to the actual skipped reference, the combined model gets simpler: pattern and scrutinee types match, and dealing with shared `&mut`s works the same way in both expressions and patterns. Trying to solve this with binding modes alone is asking too much of binding modes, and in the process pulling in stuff that would otherwise be easy to leave to deref patterns.

WaffleLapkin reviewed (commented):
Thanks a lot for working on this RFC, hopefully we'll soon be able to finally solve the papercut of not being able to deref inhereted references 🤞🏻

WaffleLapkin commented:
@nikomatsakis @joshtriplett I fail to see the benefit of "[option 1](https://github.com/rust-lang/rfcs/pull/3627#issuecomment-2113127776)". It feels like making things worse, for the ability to delay decision, even though there is only 1 semantic for `mut x = (inhereted reference)` which we plausibly would want to have in the future.

i.e. making it an error does not actually help anything, and the state where you have neither the match ergonomics way not the "explicit" way does not sound good to me.

I see how having an explicit way might be desirable, but I don't think it should be a blocker.

Jules-Bertholet commented on text/3627-match-ergonomics-2024.md:
```diff
@@ -0,0 +1,506 @@
+  references.
+- On edition ≥ 2024, `mut` on an identifier pattern does not force its binding
+  mode to by-value.
+- On edition ≥ 2024, `&` patterns can match against `&mut` references.
```
I mention this in [a footnote](https://github.com/Jules-Bertholet/rfcs/blob/match-ergonomics-2024/text/3627-match-ergonomics-2024.md#user-content-fn-nrmba-68e1807002f5d9fbe0c07b2f48675131), but `&` matching against `&mut` alleviates the downsides of the "no `ref mut` behind `&`" part of the proposal, which motivates them being included together.

Jules-Bertholet commented:
> I would still prefer to be able to consistently write a `&mut` pattern that matches an inherited reference, regardless of whether the binding mode has been converted to `ref` by an outer `&`, so long as there is a `&mut` value involved.

AIUI what you're proposing would be backward compatible. There is going to be a follow-up to this RFC at some point (to choose a syntax for mutable by-reference bindings, and possibly also to provide a solution for matching `&mut` behind `&`), so we can re-evaluate this too when the time comes.

rpjohnst commented:
The reason I'm making this proposal in the first place is because it enables us to defer "`&` matching against `&mut`," narrowing the focus of this RFC.

tmandry commented on text/3627-match-ergonomics-2024.md:
```diff
@@ -0,0 +1,506 @@
+//! Edition ≥ 2024: works with or without this rule (alternatives to above)
+// No need to abandon match ergonomics
+let &[[&a]] = &[&mut [42]]; // x: i32
+let &[&mut [&a]] = &[&mut [42]]; // x: i32
```
I think this should be

```suggestion
let &[&mut [a]] = &[&mut [42]]; // x: i32
```

tmandry commented:
@rfcbot reviewed

Having mulled it over, I think this RFC strikes the right balance on all the questions we need to decide now. I would also like to see us move forward with an implementation in nightly _as quickly as possible_ to give enough time for experimentation with the new rules and with the migration.

> Never setting default binding mode to `ref mut` behind &
> 
> ```rust
> //! All editions: works only with this rule
> let &[[a]] = &[&mut [42]]; // x: &i32
> ```
> 
> ```rust
> //! Edition ≥ 2024: works only without this rule
> let &[[&mut a]] = &[&mut [42]]; // x: i32
> ```

I would rather have the first option work than the second, especially when considering the alternatives (to binding `x: &i32` and `x: i32`, respectively) described in the RFC.

> `mut` on a binding with a by-reference binding mode is an error.

While I can see that we might want to allow this in the future, it's not entirely clear to me that this is the case. As the RFC states, "there is not much use for mutable by-reference bindings". They seem rare enough that I would guess a significant percentage of the time, they are not actually what the user wants when written, and when used correctly are unclear to the reader.

Jules-Bertholet commented:
> I would also like to see us move forward with an implementation in nightly quickly

ICYMI: https://github.com/rust-lang/rust/issues/123076

Everything in this RFC is either already in nightly under an experimental feature gate, or waiting on PR review.

Jules-Bertholet commented:
> I would rather have the first option work than the second, especially when considering the alternatives (to binding `x: &i32` and `x: i32`, respectively) described in the RFC.

Reviewing this section again, I've realized that I grossly overstated my case. For the `let &[[a]] = &[&mut [42]]; // x: &i32` example that works only with the proposed rule, `let [[a]] = &[&mut [42]];` would work just as well, whether or not we adopt the rule. I apologize for the misrepresentation, and I've replaced the example with a better one that is more fair.

rpjohnst commented:
> I would rather have the first option work than the second, especially when considering the alternatives (to binding `x: &i32` and `x: i32`, respectively) described in the RFC.

As Jules noted, it should be backwards compatible with the RFC to accept the `let &[[&mut a]] = &[&mut [42]]; // a: i32` example in the future as well. So I don't think we are fundamentally choosing between these `x: &i32` and `x: i32` examples.

Rather, given that we want to make `let &[[a]] = &[&mut [42]]; // a: &i32` work to match `let [[a]] = &[&mut [42]]; // a: &i32`, our choice is how to match these "downgraded" inherited references when switching back to the `move` binding mode:

* The RFC proposes `let &[[&a]] = &[&mut [42]]; // a: i32`, and then suggests that we allow `&` to match non-inherited `&mut`s as well: `let a: &u8 = &mut 42;`.
* We could alternatively support `let &[[&mut a]] = &[&mut [42]]; // a: i32` immediately, and leave "`&` patterns matching against `&mut`" (both inherited and non-inherited) as the backwards-compatible future extension.

joshtriplett commented:
@rfcbot resolve switch-to-option-1
@rfcbot reviewed

nikomatsakis commented:
@rfcbot resolve leaving-us-in-an-inconsitent-state

tmandry commented:
> Everything in this RFC, including the migration lint, is either already in nightly under an experimental feature gate, or waiting on PR review.

Fantastic, thanks for your efforts here. Ideally I would like to see this go in as part of the unstable 2024 edition once this RFC lands, i.e. not blocking on a stabilization FCP before that happens. I'm not sure if that's consistent or not with what we've done in the past.

> Reviewing this section again, I've realized that I grossly overstated my case. For the `let &[[a]] = &[&mut [42]]; // x: &i32` example that works only with the proposed rule, `let [[a]] = &[&mut [42]];` would work just as well, whether or not we adopt the rule.

Thanks for the correction! The new examples feel more like a wash to me, to the point where I don't know which one to prefer.

The RFC lists arguments in favor of this rule, but what are the drawbacks? I guess this one? (It sounds like it applies, even though the surrounding conversation seems to be about the `&` matching `&mut` rule, but maybe I misunderstand):

> I would still prefer to be able to consistently write a `&mut` pattern that matches an inherited reference, regardless of whether the binding mode has been converted to `ref` by an outer `&`, so long as there is a `&mut` value involved.

Are there others?

@rfcbot concern what tradeoff are we making with the "`ref mut` behind `&`" rule

If we have a clear idea of the tradeoff we're making I would think we can resolve this in a triage meeting, so nominating for discussion now.

@rustbot label I-lang-nominated

Jules-Bertholet commented:
> The RFC lists arguments in favor of this rule, but what are the drawbacks? I guess this one? (It sounds like it applies, even though the surrounding conversation seems to be about the `&` matching `&mut` rule, but maybe I misunderstand):
> 
> > I would still prefer to be able to consistently write a `&mut` pattern that matches an inherited reference, regardless of whether the binding mode has been converted to `ref` by an outer `&`, so long as there is a `&mut` value involved.

That's basically it.

```rust
//! Edition ≥ 2024
let [[&mut x]] = [&mut [42]]; // This works
// add `&`
let &[[&mut x]] = &[&mut [42]]; // This is broken by the proposed rule (example given in RFC)
```

Note that with `&` matching `&mut`, you can write:

```rust
//! Edition ≥ 2024
let [[&x]] = [&mut [42]]; // This works, with same meaning as example above
// add `&`
let &[[&x]] = &[&mut [42]]; // This also works
```

That's why `&` matching `&mut` was brought up in the conversation—it alleviates the drawbacks of the rule.

tmandry commented:
> Rather, given that we want to make `let &[[a]] = &[&mut [42]]; // a: &i32` work to match `let [[a]] = &[&mut [42]]; // a: &i32`, our choice is how to match these "downgraded" inherited references when switching back to the `move` binding mode:
> 
> * The RFC proposes `let &[[&a]] = &[&mut [42]]; // a: i32`, and then suggests that we allow `&` to match non-inherited `&mut`s as well: `let a: &u8 = &mut 42;`.
> * We could alternatively support `let &[[&mut a]] = &[&mut [42]]; // a: i32` immediately, and leave "`&` patterns matching against `&mut`" (both inherited and non-inherited) as the backwards-compatible future extension.

As I understand it, the main drawback of this would be that the compiler would no longer "do the right thing" for you when matching by reference to x in `&&mut x` (or similar). In addition to the arguments presented in the RFC.

So we are not converging on the same place: fundamentally we have to choose, or (perhaps) go for an even more conservative "non-choice" option.

@rpjohnst Are you making an argument in favor of going with the other option? If so, is there a chance you could write that up in a self-contained hackmd? Even if we accept the current RFC as-is, it would be good to include as part of the RFC alternatives section.

Jules-Bertholet commented:
> @rpjohnst Are you making an argument in favor of going with the alternative option? If so, is there a chance you could write that up in a more self-contained hackmd? Even if we accept the current RFC as-is, it would be good to include as part of the RFC alternatives section.

My understanding of @rpjohnst's position is that he thinks we should accept the union of what the two alternatives accept (is, be maximally permissive).

rpjohnst commented:
> As I understand it, the main drawback of this alternative would be that the compiler would no longer "do the right thing" for you when matching by reference to x in `&&mut x` (or similar).

I'm not proposing that we keep the old behavior of an unusable `ref mut` binding mode, as in `let &[a] = &&mut [42]; // ERROR because a: &mut i32`. Making this work with `a: &i32` was the "given that we want..." at the start of what you quoted, @tmandry. I agree with the RFC that the "binding mode is never set to `ref mut` behind a `&` value" change is useful and needs to be made over an edition.

I'm also not proposing that we be maximally permissive, @Jules-Bertholet. My primary reason for jumping into this thread has always been to make the argument that we *not* make the "`&` patterns match against `&mut` values" change (yet), because that overlaps with ongoing deref patterns work, and would be a backwards-compatible change if it turns out to be the right choice there.

However, as currently specified, the "no `ref mut` behind `&`" change interacts in an unfortunate way with the "inherited references" change: we "do the right thing" under all `&`s by using a `ref` binding mode instead of a `ref mut` binding mode, but at the same time that binding mode changes which patterns (`&` and/or `&mut`) can match the inherited reference, for when you're *not* binding by reference.

This is why the RFC to accepts `let &[[&a]] = &[&mut [42]];` but rejects `let &[[&mut a]] = &[&mut [42]];`, despite otherwise accepting things like `let [&mut a] = &mut [42];`. The RFC points to this as a supporting argument for the "`&` patterns match against `&mut` values," which would reduce the need to deal with this by supporting `&` in more places.

I look at this consequence of "no `ref mut` behind `&`" in the same way as the "`&` patterns match against `&mut` values" change. They both lead to `&` patterns in places where the (actual or inherited) scrutinee is of `&mut` type, which I would like to avoid altogether. Of course, doing this without giving up on `let &[a] = &&mut [42]; // a: &i32` means changing how "no `ref mut` under `&`" works.

So in the spirit of describing this in a self-contained way, I am proposing these changes to parts 2-4 of the RFC's reference section:
* Remove "`&` patterns can match against `&mut` references."
* `&` patterns only reset the `ref` binding mode, and `&mut` patterns only reset the `ref mut` binding mode.
* The binding mode is set to `ref` when skipping a `&` and to `ref mut` when skipping a `&mut`, regardless of the previous binding mode or whether we are behind a `&`. At the same time, the type of a binding is now determined from the *combination* of the current mode and the presence of an outer `&`.

This last point should behave the same way as the RFC (e.g. `let &[a] = &&mut [42]; // a: &i32`), *except* that it now rejects `let &[[&a]] = &[&mut [42]];` and accepts `let &[[&mut a]] = &[&mut [42]];`. I believe this makes the RFC smaller overall, leaving the meaning of these new `&` patterns up to the deref pattern work, while still solving all the original problems that led to this RFC.

Jules-Bertholet commented:
To be clear, the interaction with `&` matching against `&mut` is one of the justifications for the "no `ref mut` behind `&`" rule, but it is not the *only* one. See the ["desirable property'](https://github.com/Jules-Bertholet/rfcs/blob/match-ergonomics-2024/text/3627-match-ergonomics-2024.md#ensures-that-a-desirable-property-is-preserved) subsection in the RFC; @rpjohnst's proposal would break it irrevocably.

tmandry commented:
Many thanks for the clarifications, @rpjohnst and @Jules-Bertholet. I think I have a good grasp of your positions now.

> My primary reason for jumping into this thread has always been to make the argument that we _not_ make the "`&` patterns match against `&mut` values" change (yet), because that overlaps with ongoing deref patterns work, and would be a backwards-compatible change if it turns out to be the right choice there.

Leaving more design space open for deref patterns is desirable. At the same time, I struggle somewhat to come up with a scenario in which adopting "`&` matches `&mut`" now restricts us meaningfully in the future.

While I'd like to avoid going too deep into hypotheticals, is there any way you could sketch out the _kind of question_ this might restrict our choices on in the future, @rpjohnst?

cc @WaffleLapkin, who was also worried about interactions here.

> This last point should behave the same way as the RFC (e.g. `let &[a] = &&mut [42]; // a: &i32`), _except_ that it now rejects `let &[[&a]] = &[&mut [42]];` and accepts `let &[[&mut a]] = &[&mut [42]];`

For my own understanding, is there a "nuclear option" that rejects _both_ of these and allows us to choose one later? Does that mean delaying "`&` and `&mut` can match against inherited references" entirely, and is that something we can even do?

Jules-Bertholet commented:
> For my own understanding, is there a "nuclear option" that rejects _both_ of these and allows us to choose one later? Does that mean delaying "`&` and `&mut` can match against inherited references" entirely,

It's *possible* to do, we would have to forbid matching against inherited references in the specific situation that the rule affects. It would be confusing and irritating for users, though. Unless there a compelling reason to think we might gain new information in the future to better inform this choice (and I don't see us), delay will not do any good IMO.

rpjohnst commented:
> Leaving more design space open for deref patterns is desirable. At the same time, I struggle somewhat to come up with a scenario in which adopting "`&` matches `&mut`" now restricts us meaningfully in the future.
>
> While I'd like to avoid going too deep into hypotheticals, is there any way you could sketch out the kind of question this might restrict our choices on in the future, @rpjohnst?

It's looking like deref patterns will introduce some syntax for "deref a scrutinee of any pointer type, and then match on the result." That syntax may or may not be `&` itself. From the wider perspective of deref patterns, we might want these "universal `&` patterns" to behave in a subtly different way than whatever we come up with here, where we're only considering `&` and `&mut`.

Similarly, if we introduce "`&` matches `&mut`" now, but deref patterns choose a different syntax, we will have two different ways to match on "some kind of pointer but I don't care which." Another of the RFC's justifications for this extension of `&` is that it "makes refactoring less painful" - this is likely to apply to whatever deref patterns comes up with as well. It might be preferable to have only one language feature playing this role.

---

Regarding the "nuclear option," we would only need to reject matching on inherited references which are both A) inherited from a `&mut` value and B) behind `&`, rather than all inherited references. While rejecting `&` leaves space open for deref patterns, simultaneously rejecting `&mut` only really leaves space open to figure out the "desirable property:"

>  An `&mut` pattern is accepted if and only if removing the pattern would allow obtaining an `&mut` value.

Personally, I don't believe this is important. The language is full of ways to "talk about" something without being able to "get your hands on it," but these are enforced by the borrow checker, rather than by adjusting types ahead of time.

Consider that, given `x: &&mut i32`, the type of `*x` is still `&mut i32` and not `&i32`. Imagine that `&` and `&mut` had distinct deref operators `.d` and `.dm`, like they have distinct patterns - it would still make sense to accept `&x.d.dm` rather than `&x.d.d` or `x.d`, and to reject `&mut x.d.dm` in the borrow checker rather than the type checker.

When you rely on auto(de)ref to get a `&i32` out of `*x`, it works by desugaring to `&**x`. In fact, even when you try to get a `&mut i32` out of `*x`, the error you get is a [failure to reborrow `**x`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e34c1db383017192e8c2f59e3e5f74e4).

To me it seems *more* in line with the rest of the language to accepting `&mut` patterns (the pattern corresponding to my made-up `.dm` expressions) even in places that don't permit mutation, or mutable reborrowing, or otherwise materializing a `&mut`.

tmandry commented:
The deciding factor is ultimately whether the lang team has _considered_ all of the information it thinks is relevant, or will have time to do so, without either delaying the edition or risking this RFC as part of it. 

> > For my own understanding, is there a "nuclear option" that rejects _both_ of these and allows us to choose one later? Does that mean delaying "`&` and `&mut` can match against inherited references" entirely,
> 
> It's *possible* to do, we would have to forbid matching against inherited references in the specific situation that the rule affects. It would be confusing and irritating for users, though. 

What about the second part of my question — delaying "`&` and `&mut` can match against inherited references" until we have a firm decision one way or the other?

We all agree this feature is desirable, and I would like to ship it as part of the story we tell for the edition. But we also have to prioritize decisions that _must_ be made for the edition (and we are whittling those down quickly, knock on wood).

Is this feature inherently edition sensitive on its own?

Jules-Bertholet commented:
> Is this feature inherently edition sensitive on its own?

Yes, at least if we adopt the "eat-one-layer" model as proposed.

Jules-Bertholet commented:
I've incorporated @traviscross's suggested additions from the 2024-06-26 T-lang design meeting ([minutes](https://hackmd.io/_ey51TmXRqyHq1fQ_t9pmQ)). We now allow `let [&mut x] = &[&mut 42]`. In addition, “`&` matches `&mut`” now works on all editions.

tmandry commented:
Thank you for the extensive work you put into this RFC, @Jules-Bertholet, and to @traviscross for framing and driving consensus on the remaining details. I'm very impressed by the attention to detail you put into the work and by how well you collaborated with one another and with @Nadrieril to get it over the line. I'm extremely happy with the result and with how confident I feel in it now. 

It looks like we have consensus on the rules as put forth in this RFC. 

@rfcbot resolve what tradeoff are we making with the "`ref mut` behind `&`" rule

Part of the rationale of accepting this RFC as it is has now become the convergence of several different lines of reasoning: The one put forth in this RFC, the one put forth by @traviscross in the most recent lang design meeting on the topic, and the design axioms shared by @Nadrieril in Zulip after the meeting. I would strongly suggest including links to the latter two in the RFC itself, because not only do they strengthen its case, they provide alternative framings for understanding both the problem space of the proposal and the mindset of those around the lang team when it was accepted.

WaffleLapkin commented on text/3627-match-ergonomics-2024.md:
```diff
@@ -0,0 +1,519 @@
+  references.
+- On edition ≥ 2024, `mut` on an identifier pattern does not force its binding
+  mode to by-value.
+- On all editions, `&` patterns can match against `&mut` references.
```
This is a breaking change.

Rust uses the patterns for type inference, so changing what `&` can match will/should will stop some code from compiling. A simple example is

```rust
let &_ = generic_return(); // the function is inferred to return `&_`
```
([full working example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0e8440c88322611b90ed2f06fa999167))

Jules-Bertholet replied:
I hadn’t considered that case, but it actually *does* work under the current implementation: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=2acaaa1d97f85b4f2cb04f5160ee5848)

I’ll add tests and update the RFC, but for now I don’t think there is real breakage.

Jules-Bertholet replied:
Ahh, looking at the implementation I think I see what is happening. [This](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=1acce770a0ed215f7315679c5437016f) doesn’t work

Jules-Bertholet replied:
I’ve updated the RFC.

WaffleLapkin replied:
> However, if the type of the scrutinee is unknown, an `&` pattern will still constrain inference to force it to be a shared reference. (97bf386e8e04e8267b706460602fb08046286f87)

I don't think this is right?

In the sense that while this is the currently implemented behavior, it's not a very good behavior. I think the inference part should be in tune/agreement with the matching part. If `&` can match `&mut`, it shouldn't guide inference towards using `&`.

Nadrieril replied:
Match ergonomics interact poorly with inference everywhere: `let Foo(x) = function();` guides inference to get a `Foo` out of `function()` even when it could equally get a `&Foo`. I've often had spurious "type X is not Copy" errors because of a type error somewhere else that made inference ignore match ergonomics in this way.

Jules-Bertholet replied:
_Ideally_, we would constrain to `&` or `&mut` with priority for the former. But I don't think the compiler is smart enough for that

nikomatsakis commented:
Thanks to everyone for this, I'm really looking forward to this change.

On Sun, Jul 14, 2024, at 2:34 AM, Travis Cross wrote:
> 
> 
> The lang team has accepted this RFC and it's now been *merged*.
> 
> Thanks to @Jules-Bertholet <https://github.com/Jules-Bertholet> for pushing this important work forward, and thanks to @Nadrieril <https://github.com/Nadrieril> for earlier work here.
> 
> For further updates, follow the tracking issue:
> 
>  • rust-lang/rust#123076 <https://github.com/rust-lang/rust/issues/123076>
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/rust-lang/rfcs/pull/3627#issuecomment-2227213469>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AABF4ZQ3QZ5E4KGO6QEL7Y3ZMILVRAVCNFSM6AAAAABHJXJXDGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDEMRXGIYTGNBWHE>.
> You are receiving this because you were mentioned.Message ID: ***@***.***>
>
